File: iOS/SnowTeeth/Utilities/AppPreferences.swift
//
//  AppPreferences.swift
//  SnowTeeth
//
//  Created by SnowTeeth
//

import Foundation
import Combine

class AppPreferences: ObservableObject {
    private let defaults = UserDefaults.standard

    // Keys
    private enum Keys {
        static let downhillHard = "downhill_hard"
        static let downhillMedium = "downhill_medium"
        static let downhillEasy = "downhill_easy"
        static let uphillEasy = "uphill_easy"
        static let uphillMedium = "uphill_medium"
        static let uphillHard = "uphill_hard"
        static let useMetric = "use_metric"
        static let visualizationStyle = "visualization_style"
        static let isTracking = "is_tracking"
        static let useAlternateAscentVisualization = "use_alternate_ascent_visualization"
    }

    // Default values (in mph)
    static let defaultDownhillHard: Float = 10.0
    static let defaultDownhillMedium: Float = 6.0
    static let defaultDownhillEasy: Float = 2.0
    static let defaultUphillEasy: Float = 1.0
    static let defaultUphillMedium: Float = 3.0
    static let defaultUphillHard: Float = 5.0

    // Threshold properties
    @Published var downhillHardThreshold: Float {
        didSet { defaults.set(downhillHardThreshold, forKey: Keys.downhillHard) }
    }

    @Published var downhillMediumThreshold: Float {
        didSet { defaults.set(downhillMediumThreshold, forKey: Keys.downhillMedium) }
    }

    @Published var downhillEasyThreshold: Float {
        didSet { defaults.set(downhillEasyThreshold, forKey: Keys.downhillEasy) }
    }

    @Published var uphillEasyThreshold: Float {
        didSet { defaults.set(uphillEasyThreshold, forKey: Keys.uphillEasy) }
    }

    @Published var uphillMediumThreshold: Float {
        didSet { defaults.set(uphillMediumThreshold, forKey: Keys.uphillMedium) }
    }

    @Published var uphillHardThreshold: Float {
        didSet { defaults.set(uphillHardThreshold, forKey: Keys.uphillHard) }
    }

    // Settings properties
    @Published var useMetric: Bool {
        didSet { defaults.set(useMetric, forKey: Keys.useMetric) }
    }

    @Published var visualizationStyle: VisualizationStyle {
        didSet { defaults.set(visualizationStyle.rawValue, forKey: Keys.visualizationStyle) }
    }

    @Published var isTracking: Bool {
        didSet { defaults.set(isTracking, forKey: Keys.isTracking) }
    }

    @Published var useAlternateAscentVisualization: Bool {
        didSet { defaults.set(useAlternateAscentVisualization, forKey: Keys.useAlternateAscentVisualization) }
    }

    init() {
        // Load saved values or use defaults
        self.downhillHardThreshold = defaults.object(forKey: Keys.downhillHard) as? Float ?? Self.defaultDownhillHard
        self.downhillMediumThreshold = defaults.object(forKey: Keys.downhillMedium) as? Float ?? Self.defaultDownhillMedium
        self.downhillEasyThreshold = defaults.object(forKey: Keys.downhillEasy) as? Float ?? Self.defaultDownhillEasy
        self.uphillEasyThreshold = defaults.object(forKey: Keys.uphillEasy) as? Float ?? Self.defaultUphillEasy
        self.uphillMediumThreshold = defaults.object(forKey: Keys.uphillMedium) as? Float ?? Self.defaultUphillMedium
        self.uphillHardThreshold = defaults.object(forKey: Keys.uphillHard) as? Float ?? Self.defaultUphillHard
        self.useMetric = defaults.bool(forKey: Keys.useMetric)

        let styleString = defaults.string(forKey: Keys.visualizationStyle) ?? VisualizationStyle.flame.rawValue
        self.visualizationStyle = VisualizationStyle(rawValue: styleString) ?? .flame

        self.isTracking = defaults.bool(forKey: Keys.isTracking)
        self.useAlternateAscentVisualization = defaults.bool(forKey: Keys.useAlternateAscentVisualization)
    }
}

File: android/app/src/main/java/com/snowteeth/app/util/AppPreferences.kt
package com.snowteeth.app.util

import android.content.Context
import android.content.SharedPreferences
import com.snowteeth.app.model.VisualizationStyle

class AppPreferences(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    companion object {
        private const val PREFS_NAME = "SnowTeethPrefs"

        // Threshold keys
        private const val KEY_DOWNHILL_HARD = "downhill_hard"
        private const val KEY_DOWNHILL_MEDIUM = "downhill_medium"
        private const val KEY_DOWNHILL_EASY = "downhill_easy"
        private const val KEY_UPHILL_EASY = "uphill_easy"
        private const val KEY_UPHILL_MEDIUM = "uphill_medium"
        private const val KEY_UPHILL_HARD = "uphill_hard"

        // Settings keys
        private const val KEY_USE_METRIC = "use_metric"
        private const val KEY_VIS_STYLE = "visualization_style"
        private const val KEY_TRACKING = "is_tracking"
        private const val KEY_ALTERNATE_ASCENT = "use_alternate_ascent_visualization"

        // Default values (in mph)
        const val DEFAULT_DOWNHILL_HARD = 10.0f
        const val DEFAULT_DOWNHILL_MEDIUM = 6.0f
        const val DEFAULT_DOWNHILL_EASY = 2.0f
        const val DEFAULT_UPHILL_EASY = 1.0f
        const val DEFAULT_UPHILL_MEDIUM = 3.0f
        const val DEFAULT_UPHILL_HARD = 5.0f
    }

    // Threshold getters/setters
    var downhillHardThreshold: Float
        get() = prefs.getFloat(KEY_DOWNHILL_HARD, DEFAULT_DOWNHILL_HARD)
        set(value) = prefs.edit().putFloat(KEY_DOWNHILL_HARD, value).apply()

    var downhillMediumThreshold: Float
        get() = prefs.getFloat(KEY_DOWNHILL_MEDIUM, DEFAULT_DOWNHILL_MEDIUM)
        set(value) = prefs.edit().putFloat(KEY_DOWNHILL_MEDIUM, value).apply()

    var downhillEasyThreshold: Float
        get() = prefs.getFloat(KEY_DOWNHILL_EASY, DEFAULT_DOWNHILL_EASY)
        set(value) = prefs.edit().putFloat(KEY_DOWNHILL_EASY, value).apply()

    var uphillEasyThreshold: Float
        get() = prefs.getFloat(KEY_UPHILL_EASY, DEFAULT_UPHILL_EASY)
        set(value) = prefs.edit().putFloat(KEY_UPHILL_EASY, value).apply()

    var uphillMediumThreshold: Float
        get() = prefs.getFloat(KEY_UPHILL_MEDIUM, DEFAULT_UPHILL_MEDIUM)
        set(value) = prefs.edit().putFloat(KEY_UPHILL_MEDIUM, value).apply()

    var uphillHardThreshold: Float
        get() = prefs.getFloat(KEY_UPHILL_HARD, DEFAULT_UPHILL_HARD)
        set(value) = prefs.edit().putFloat(KEY_UPHILL_HARD, value).apply()

    // Settings getters/setters
    var useMetric: Boolean
        get() = prefs.getBoolean(KEY_USE_METRIC, false)
        set(value) = prefs.edit().putBoolean(KEY_USE_METRIC, value).apply()

    var visualizationStyle: VisualizationStyle
        get() {
            val styleName = prefs.getString(KEY_VIS_STYLE, VisualizationStyle.FLAME.name)
            return VisualizationStyle.valueOf(styleName ?: VisualizationStyle.FLAME.name)
        }
        set(value) = prefs.edit().putString(KEY_VIS_STYLE, value.name).apply()

    var isTracking: Boolean
        get() = prefs.getBoolean(KEY_TRACKING, false)
        set(value) = prefs.edit().putBoolean(KEY_TRACKING, value).apply()

    var useAlternateAscentVisualization: Boolean
        get() = prefs.getBoolean(KEY_ALTERNATE_ASCENT, false)
        set(value) = prefs.edit().putBoolean(KEY_ALTERNATE_ASCENT, value).apply()
}
